src/main/perl/RuleBasedEditor.pm

```diff
@@ -0,0 +1,1023 @@
+
+used to indicate how to interpret the configuration value. It is used mainly for
+boolean values, list and hashes. See LINE_VALUE_xxx constants below for the possible values.
+
+=back 
+
+For an example of rules, look at ncm-dpmlfc or ncm-xrootd source code in
+configuration-modules-grid repository.
+
+
+=head2 Public methods
+
+=over
+
+=item new
+
+This is the constructor. It mainly executes the FileEditor constructor and supports the
+same arguments.
+
+==cut
+
+sub new
+{
+    my $class = shift;
+    my $self = $class->SUPER::new (@_);
+    return $self;
+}
+
+
+=pod
+
+=item updateFile
+
+Update configuration file contents,  applying configuration rules.
+
+Arguments :
+    config_rules: config rules corresponding to the file to build
+    config_options: configuration parameters used to build actual configuration
+    options: a hash setting options to modify the behaviour of this function
+
+Supported entries for options hash:
+    always_rules_only: if true, apply only rules with ALWAYS condition (D: false)
+    remove_if_undef: if true, remove matching configuration line is rule condition is not met (D: false)
+
+Return value
+    always 0
+    
+=cut
+
+sub updateFile {
+  my $function_name = "updateConfigFile";
+  my ($self, $config_rules, $config_options, $parser_options) = @_;
+
+  unless ( $config_rules ) {
+    *$self->{LOG}->error("$function_name: 'config_rules' argument missing (internal error)");
+    return 1;
+  }
+  unless ( $config_options ) {
+    *$self->{LOG}->error("$function_name: 'config_options' argument missing (internal error)");
+    return 1;
+  }
+  unless ( defined($parser_options) ) {
+    *$self->{LOG}->debug(2,"$function_name: 'parser_options' undefined");
+    $parser_options = {};
+  }
+
+  $self->seek_begin();
+
+  # Check that config file has an appropriate header
+  my $intro_pattern = "# This file is managed by Quattor";
+  my $intro = "# This file is managed by Quattor - DO NOT EDIT lines generated by Quattor";
+  $self->add_or_replace_lines(qr/^$intro_pattern/,
+                              qr/^$intro$/,
+                              $intro."\n#\n",
+                              BEGINNING_OF_FILE,
+                             );
+  
+  $self->_apply_rules($config_rules,
+                      $config_options,
+                      $parser_options);
+
+  return 0;
+}
+
+
+=pod
+
+=back
+
+=head2 Private methods
+
+=over
+
+=item formatAttrValue
+
+This function formats an attribute value based on the value format specified.
+
+Arguments :
+    attr_value : attribue value
+    line_fmt : line format (see LINE_FORMAT_xxx constants)
+    value_fmt : value format (see LINE_VALUE_xxx constants)
+    value_opt : value interpretation/formatting options (bitmask, see LINE_VALUE_OPT_xxx constants)
+
+=cut
+
+sub _formatAttributeValue {
+  my $function_name = "_formatAttributeValue";
+  my ($self, $attr_value, $line_fmt, $value_fmt, $value_opt) = @_;
+
+  unless ( defined($attr_value) ) {
```

Comment:
Why? Mistakes are always possible by somebody who would work on this module... Any drawback?